mrmustard/lab_dev/circuit_components.py

```diff
@@ -233,6 +233,42 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and other.wires.ket and not other.wires.bra:
+            # self has ket and bra, other has ket
+            return self @ other @ other.adjoint
+
+        if self.wires.ket and self.wires.bra and not other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has and bra
+            return self @ other @ other.adjoint
```

Comment:
would both of them can be merged?

 ``if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):``